# Shopify Summer 2022 Data Science Intern Challenge

## Question 1

In [2]:
import pandas as pd

In [3]:
data = pd.read_csv('Data2019.csv')

In [4]:
data.head()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11

In [5]:
data.describe()

order_id      shop_id      user_id   order_amount  total_items
count  5000.000000  5000.000000  5000.000000    5000.000000   5000.00000
mean   2500.500000    50.078800   849.092400    3145.128000      8.78720
std    1443.520003    29.006118    87.798982   41282.539349    116.32032
min       1.000000     1.000000   607.000000      90.000000      1.00000
25%    1250.750000    24.000000   775.000000     163.000000      1.00000
50%    2500.500000    50.000000   849.000000     284.000000      2.00000
75%    3750.250000    75.000000   925.000000     390.000000      3.00000
max    5000.000000   100.000000   999.000000  704000.000000   2000.00000

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   order_id        5000 non-null   int64 
 1   shop_id         5000 non-null   int64 
 2   user_id         5000 non-null   int64 
 3   order_amount    5000 non-null   int64 
 4   total_items     5000 non-null   int64 
 5   payment_method  5000 non-null   object
 6   created_at      5000 non-null   object
dtypes: int64(5), object(2)
memory usage: 273.6+ KB


In [7]:
for col in data:
    print(col, data[col].unique())

order_id [   1    2    3 ... 4998 4999 5000]
shop_id [ 53  92  44  18  58  87  22  64  52  66  40  54 100  42  17  28  83  63
  97  88  75  73  82  47  59  76  57   7  34  61  84  10   1  21  99  29
  33   3  81  30  79  51  80  60  86  91  72  11  14   5  20  41   8  48
  56  55  50  35  70  77  74  13  12  68  69  25   6  98  24   2  49  15
  71  16   9  78  93  37  26  43  85  19  27  31  67  36  39  89  32  23
  96  95  38  62   4  90  94  46  65  45]
user_id [746 925 861 935 883 882 915 761 914 788 848 983 799 709 849 607 731 752
 898 987 789 985 964 917 942 944 790 857 839 712 800 704 781 979 961 921
 756 793 847 934 792 759 969 954 791 714 768 984 959 968 842 823 851 739
 837 908 720 981 962 887 988 743 812 994 876 725 813 960 862 744 975 806
 838 868 817 707 711 713 998 705 835 856 810 758 850 741 972 923 740 803
 953 732 926 774 893 885 755 745 958 997 949 884 963 721 723 777 894 845
 742 760 930 855 828 906 899 715 718 808 872 986 734 795 990 877 913 941
 754 797 970 724 778 

In [8]:
for col in data:
    print(col, data[col].nunique())

order_id 5000
shop_id 100
user_id 301
order_amount 258
total_items 8
payment_method 3
created_at 4991


In [9]:
for col in data:
    print(col, data[col].value_counts())

order_id 2047    1
2608    1
4647    1
2600    1
553     1
       ..
3263    1
1218    1
3267    1
1222    1
2049    1
Name: order_id, Length: 5000, dtype: int64
shop_id 53    68
71    66
19    64
13    63
89    61
      ..
74    38
56    37
67    37
38    35
85    35
Name: shop_id, Length: 100, dtype: int64
user_id 718    28
868    27
847    26
791    26
727    25
       ..
719     9
998     9
902     9
750     7
717     7
Name: user_id, Length: 301, dtype: int64
order_amount 153    87
306    85
354    82
312    75
156    75
       ..
404     1
948     1
920     1
816     1
980     1
Name: order_amount, Length: 258, dtype: int64
total_items 2       1832
1       1830
3        941
4        293
5         77
2000      17
6          9
8          1
Name: total_items, dtype: int64
payment_method credit_card    1735
debit          1671
cash           1594
Name: payment_method, dtype: int64
created_at 2017-03-28 4:00:00     3
2017-03-07 4:00:00     2
2017-03-09 10:46:09    2
2017-03-29 7:10:18

In [27]:
items_2000 = data[(data['total_items'] == 2000)]
items_2000

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
520        521       42      607        704000         2000    credit_card   
1104      1105       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1436      1437       42      607        704000         2000    credit_card   
1562      1563       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
2153      2154       42      607        704000         2000    credit_card   
2297      2298       42      607        704000         2000    credit_card   
2835      2836       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
3332      3333       42      607        704000         2000    credit_card   
4056      4057       42      607        704000         2000    credit_card   
4646      4647       42      607        704000         2000    credit_card   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   

              created_at  
15    2017-03-07 4:00:00  
60    2017-03-04 4:00:00  
520   2017-03-02 4:00:00  
1104  2017-03-24 4:00:00  
1362  2017-03-15 4:00:00  
1436  2017-03-11 4:00:00  
1562  2017-03-19 4:00:00  
1602  2017-03-17 4:00:00  
2153  2017-03-12 4:00:00  
2297  2017-03-07 4:00:00  
2835  2017-03-28 4:00:00  
2969  2017-03-28 4:00:00  
3332  2017-03-24 4:00:00  
4056  2017-03-28 4:00:00  
4646  2017-03-02 4:00:00  
4868  2017-03-22 4:00:00  
4882  2017-03-25 4:00:00

In [28]:
items_Not_2000 = data[(data['total_items'] != 2000)]
items_Not_2000

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0            1       53      746           224            2           cash   
1            2       92      925            90            1           cash   
2            3       44      861           144            1           cash   
3            4       18      935           156            1    credit_card   
4            5       18      883           156            1    credit_card   
...        ...      ...      ...           ...          ...            ...   
4995      4996       73      993           330            2          debit   
4996      4997       48      789           234            2           cash   
4997      4998       56      867           351            3           cash   
4998      4999       60      825           354            2    credit_card   
4999      5000       44      734           288            2          debit   

               created_at  
0     2017-03-13 12:36:56  
1     2017-03-03 17:38:52  
2      2017-03-14 4:23:56  
3     2017-03-26 12:43:37  
4      2017-03-01 4:35:11  
...                   ...  
4995  2017-03-30 13:47:17  
4996  2017-03-16 20:36:16  
4997   2017-03-19 5:42:42  
4998  2017-03-16 14:51:18  
4999  2017-03-18 15:48:18  

[4983 rows x 7 columns]

In [29]:
items_Not_2000['order_amount'].describe()

count      4983.000000
mean        754.091913
std        5314.092293
min          90.000000
25%         163.000000
50%         284.000000
75%         390.000000
max      154350.000000
Name: order_amount, dtype: float64

In [31]:
data['order_amount'].describe()

count      5000.000000
mean       3145.128000
std       41282.539349
min          90.000000
25%         163.000000
50%         284.000000
75%         390.000000
max      704000.000000
Name: order_amount, dtype: float64

## a. Problem with Average Order Value (AOV):
The Average Order value is sensitive to skewed data. In our data, a customer with user_id = 607 made 2000 purchases of worth 704000 exactly 17 times, resulting in an incorrectly inflated Average Order Value.

## b. Alternate Metric:
Alternatively we can use Median which is less sensitive to extreme values.

## c. Value:
Median of order amount is $284.0.

## Question 2:
For this question you’ll need to use SQL. Follow this link to access the data set required for the challenge. Please use queries to answer the following questions. Paste your queries along with your final numerical answers below.

## a.How many orders were shipped by Speedy Express in total?

SELECT count(*) <br>
FROM Orders <br>
WHERE ShipperID = 1

### Value: 54

## b. What is the last name of the employee with the most orders?

SELECT LastName <br>
FROM (SELECT LastName, MAX(total) <br>
FROM(SELECT e.LastName, count(o.EmployeeID)total FROM Employees AS e, Orders AS o <br>
WHERE e.EmployeeID = o.EmployeeID <br>
GROUP BY o.EmployeeID, e.LastName));

### Employee's Last Name: Peacock

## c. What product was ordered the most by customers in Germany?

SELECT ProductName <br>
FROM (SELECT ProductName, MAX(MostOrdered) <br>
FROM (SELECT p.ProductName, count(*) MostOrdered FROM Customers AS c, Orders As o, OrderDetails As od, Products As p <br>
WHERE c.Country = 'Germany'and c.CustomerID = o.CustomerID and o.OrderID = od.OrderID and od.ProductID = p.ProductID <br>
GROUP BY p.ProductName));

### Product Name: Gorgonzola Telino